### PR TITLE
Sundial on AWS Batch blog post

### PR DESCRIPTION
**Pull request by kevinoriordan**

### CODE REVIEW

File: _posts/2017-08-04-sundial-batch.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+- tech
+- personalization
+---
+
+# Sundial on AWS Batch
+
+
+About a year ago, we open sourced Sundial (https://github.com/gilt/sundial) , a batch job orchestration system built on top of Amazon EC2 Container Service.
+
+We built Sundial to provide the following features on top of the standard ECS setup:
```
Comment (nemo83) [inline _posts/2017-08-04-sundial-batch.md]: leveraging instead of on top (that you just used above)

File: _posts/2017-08-04-sundial-batch.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+ - Streaming Logs (to Cloudwatch and S3 and live in Sundial UI)
+ - Metadata collection (through Graphite and displayed live in Sundial UI)
+ - Dependency management between jobs
+ - Retry strategies for failed jobs
+ - Cron style scheduling for jobs
+ - Email status reporting for jobs
+ - Pagerduty integration for notifying team members about failing critical jobs
+
+ Other solutions available at the time didn't suit our needs. Solutions we considered included Chronos https://mesos.github.io/chronos/ which lacked the features we needed and required a Mesos cluster, and Airbnb Airflow, which was immature at the time. http://airbnb.io/projects/airflow/
```
Comment (nemo83) [inline _posts/2017-08-04-sundial-batch.md]: [Chronos](https://mesos.github.io/chronos/)
[Airbnb Airflow](http://airbnb.io/projects/airflow/)

File: _posts/2017-08-04-sundial-batch.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+### What we love about Batch
+
+  - Managed compute environment. This means AWS handles scaling up and down the cluster in response to workload.
+  - Heterogenous instance types (useful when we have outlier jobs taking large amounts of CPU/memory resources)
+  - Spot instances (save over half on on-demand instance costs)
+  - Easy integration with Cloudwatch Logs (stdout and stderr captured automatically)
+
+### What sucks
```
Comment (nemo83) [inline _posts/2017-08-04-sundial-batch.md]: <3

File: _posts/2017-08-04-sundial-batch.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ - Easy transition, processes can be a mixed workload of jobs running on ECS and Batch.
+ - Configurable backoff strategy for job retries.
+ - Nice dashboard of processes (At a glance see what's green and what's red)
+
+ Sure enough the above can be configured through hooking up lambdas/SNS messages etc. but Sundial gives it to you out of the box.
+
+ ### What next?
+
+ Although Sundial on Batch solves the immediate needs for us (The Gilt/HBC Personalization team), maintaining a job system counts as technical debt that
+ is a distraction from product focused tasks. The HBC data team who have very similar requirements to us have started investigating making Airflow (by Airbnb) run on Batch so this is a solution we may consider in the future.
```
Comment (nemo83) [inline _posts/2017-08-04-sundial-batch.md]: This sounds like you're not really happy with Sundial + Batch. Is what you really meant here?

I'm not familiar with Airflow, but imho Sundial + Batch = <3.

Comment (kevinoriordan) [inline _posts/2017-08-04-sundial-batch.md]: Not so much that we're unhappy but talking to Oliver, Airflow offers a better experience UI wise and would mean one less service for us to maintain. Data team are migrating away from Sundial to Airflow anyway. Oliver said they'd give us a demo next week.

File: _posts/2017-08-04-sundial-batch.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ - Email status reporting for jobs
+ - Pagerduty integration for notifying team members about failing critical jobs
+
+ Other solutions available at the time didn't suit our needs. Solutions we considered included [Chronos](https://mesos.github.io/chronos/) which lacked the features we needed and required a Mesos cluster, and [Airbnb Airflow](http://airbnb.io/projects/airflow/), which was immature at the time.
+
+ At the time, we chose ECS because we hoped to take advantages of AWS features such as autoscaling in order to save costs
+ by scaling the cluster up and down by demand. In practice, this required too much manual effort and moving parts so we lived with a long running cluster
+ scaled to handle peak load.
+
+ Since then our needs have grown and we have jobs ranging in size from a couple of hundred MB of memory to 60GB of memory. Having a cluster scaled
```
Comment (JasonTam) [inline _posts/2017-08-04-sundial-batch.md]: Since then `,`

File: _posts/2017-08-04-sundial-batch.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ to handle peak load with all these job sizes has become too expensive. Most job failure noise has been due to cluster resources not being available or smaller jobs taking up space on instances meant to be dedicated to bigger jobs. (ECS is weak when it comes to task placement strategies).
```
Comment (JasonTam) [inline _posts/2017-08-04-sundial-batch.md]: has become too expensive -> `had` (if we've already moved to batch)

File: _posts/2017-08-04-sundial-batch.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+ Thankfully AWS have come along with their own enhancements on top of ECS in the form of AWS Batch.
```
Comment (JasonTam) [inline _posts/2017-08-04-sundial-batch.md]: can link AWS batch if you want

File: _posts/2017-08-04-sundial-batch.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+ - Streaming Logs (to Cloudwatch and S3 and live in Sundial UI)
+ - Metadata collection (through Graphite and displayed live in Sundial UI)
+ - Dependency management between jobs
+ - Retry strategies for failed jobs
+ - Cron style scheduling for jobs
+ - Email status reporting for jobs
+ - Pagerduty integration for notifying team members about failing critical jobs
+
+ Other solutions available at the time didn't suit our needs. Solutions we considered included [Chronos](https://mesos.github.io/chronos/) which lacked the features we needed and required a Mesos cluster, and [Airbnb Airflow](http://airbnb.io/projects/airflow/), which was immature at the time.
```
Comment (JasonTam) [inline _posts/2017-08-04-sundial-batch.md]: Curious:
Was spotify's luigi ever considered?

File: _posts/2017-08-04-sundial-batch.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+Sundial still provides features that Batch doesn't provide:
+ - Email reporting
+ - Pagerduty integration
+ - Easy transition, processes can be a mixed workload of jobs running on ECS and Batch.
+ - Configurable backoff strategy for job retries.
+ - Time limits for jobs. If a job hangs, we can kill and retry after a certain period of time
+ - Nice dashboard of processes (At a glance see what's green and what's red)
+
+ Sure enough some of the above can be configured through hooking up lambdas/SNS messages etc. but Sundial gives it to you out of the box.
```
Comment (JasonTam) [inline _posts/2017-08-04-sundial-batch.md]: Sure enough `,`

File: _posts/2017-08-04-sundial-batch.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+ ### What next?
+
+ Although Sundial on Batch solves the immediate needs for us (The Gilt/HBC Personalization team), maintaining a job system counts as technical debt that
+ is a distraction from product focused tasks. The HBC data team who have very similar requirements to us have started investigating making Airflow (by Airbnb) run on Batch so this is a solution we may consider in the future.
```
Comment (JasonTam) [inline _posts/2017-08-04-sundial-batch.md]: The HBC data team`,` who have very similar requirements to us`,` have started investigating making Airflow (by Airbnb) run on Batch`;` so this is a solution we may consider in the future.

File: _posts/2017-08-04-sundial-batch.md (hunk 4 above)
Comment (JasonTam) [inline _posts/2017-08-04-sundial-batch.md]: It's funny cuz I think spotify migrated from their homegrown luigi project to using airflow as well.

File: _posts/2017-08-04-sundial-batch.md (hunk 8 above)
Comment (kevinoriordan) [inline _posts/2017-08-04-sundial-batch.md]: Don't recall it being considered but that's probably a question for Zach (I think he still reads these PRs ;) ).

**Pull request merged**